subcommands/devices/list.go

```diff
@@ -14,29 +17,111 @@ import (
 
 var (
 	deviceNoShared      bool
-	deviceNoOwner       bool
 	deviceByTag         string
 	deviceByFactory     string
 	deviceByGroup       string
 	deviceInactiveHours int
 	deviceUuid          string
+	showColumns         []string
 )
 
+type column struct {
+	FieldName string
+	Formatter func(d *client.Device, fieldName string) string
+}
+
+func defFormatter(d *client.Device, fieldName string) string {
+	e := reflect.ValueOf(d).Elem()
+	val := e.FieldByName(fieldName).Interface()
+	if reflect.TypeOf(val).String() == "[]string" {
+		return strings.Join(val.([]string), ",")
+	}
+	return fmt.Sprintf("%v", val)
+}
+
+func statusFormatter(d *client.Device, fieldName string) string {
+	status := "OK"
+	if len(d.Status) > 0 {
+		status = d.Status
+	}
+	if len(d.LastSeen) > 0 && !d.Online(deviceInactiveHours) {
+		status = "OFFLINE"
+	}
+	return status
+}
+
+var ownerCache = make(map[string]string)
+
+func ownerFormatter(d *client.Device, fieldName string) string {
+	name, ok := ownerCache[d.Owner]
+	if ok {
+		return name
+	}
+	logrus.Debugf("Looking up user %s in factory %s", d.Owner, d.Factory)
+	users, err := api.UsersList(d.Factory)
+	if err != nil {
+		logrus.Errorf("Unable to look up users: %s", err)
+		return "???"
+	}
+	id := "<not in factory: " + d.Factory + ">"
+	for _, user := range users {
+		ownerCache[user.PolisId] = user.Name
+		if user.PolisId == d.Owner {
+			id = user.Name
+		}
+	}
+	return id
+}
+
+func groupFormatter(d *client.Device, fieldName string) string {
+	if d.Group != nil {
+		return d.Group.Name
+	}
+	return ""
+}
+
+var Columns = map[string]column{
+	"name":           {FieldName: "Name", Formatter: defFormatter},
```

Comment:
I'm a fan of Python reflection, but for Golang I'd prefer to find a way to solve a problem without it.
There are many pitfalls in Golang reflection e.g. there is no type checking and all sorts of mistakes would compile well, but panic later on.

One idea, which is almost as neat as what you've done is this:
```
type formatter func *client.Device) string
var Columns = map[string]formatter{
  "name": func (d *client.Device) string {return defFormatter(d.Name)},
  ...
}
```

Reply:
Note that you might save a some characters (if you care) with a simple type alias `type Device = client.Device`.

Reply:
Forgot to mention that this way a default formatter might accept an interface and do simple type assertion instead of reflection i.e. it becomes simpler.
At the same time more specialized formatters would be able to do a proper type checking via a function signature at compile time.

Reply:
that's a better idea. i'll try and re-work it with that approach.

Reply:
i just did a force-push cleanup. let me know what you think.

Reply:
Looks awesome